### PR TITLE
C#: Fix DB inconsistencies with extends/2

### PR DESCRIPTION
**Pull request by calumgrant**
When a base class is unknown (due to a missing reference), Roslyn will incorrectly give a base class  of `Object`, which gets written to the `extends/2` table. This can cause DB inconsistencies. Instead, do not populate `Object` as a base class, and instead deduce it in QL.

### CODE REVIEW

File: csharp/ql/src/semmle/code/csharp/Type.qll [hunk 1]
```diff
@@ -111,7 +111,14 @@ class ValueOrRefType extends DotNet::ValueOrRefType, Type, Attributable, @value_
...
 
   /** Gets the immediate base class of this class, if any. */
-  Class getBaseClass() { extend(this, getTypeRef(result)) }
+  Class getBaseClass() {
+    if extend(this, _)
+    then extend(this, getTypeRef(result))
+    else (
+      not this instanceof ObjectType and
+      result instanceof ObjectType
+    )
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/Type.qll]: I would write
```
extend(this, getTypeRef(result))
or
not extend(this, _) and
not this instanceof ObjectType and
result instanceof ObjectType
```

File: csharp/ql/src/semmle/code/csharp/Type.qll [hunk 2]
```diff
@@ -111,7 +111,13 @@ class ValueOrRefType extends DotNet::ValueOrRefType, Type, Attributable, @value_
   }
 
   /** Gets the immediate base class of this class, if any. */
-  Class getBaseClass() { extend(this, getTypeRef(result)) }
+  Class getBaseClass() {
+    extend(this, getTypeRef(result))
+    or
+    not extend(this, _) and
+    not this instanceof ObjectType and
+    result instanceof ObjectType
```
Comment (tamasvajk) [inline csharp/ql/src/semmle/code/csharp/Type.qll:119]: Why is this `result instanceof ObjectType` and not `result = ObjectType`?

Comment (hvitved) [reply csharp/ql/src/semmle/code/csharp/Type.qll:119]: The latter is not valid QL syntax, as the right-hand-side is a QL class and not an expression. `result = any(ObjectType ot)` would work, but `instanceof` is better IMO. Note that the extractor will guarantee that there is indeed only one element inside `ObjectType`.

**Comment by hvitved** [comment]
I had to make a similar change on https://github.com/github/codeql/pull/4017.

**Pull request closed without merging**